### PR TITLE
Use completion item InsertionText in the debugger, even before any non-debugger completion sessions have run

### PR DESCRIPTION
**Pull request by dpoeschl**
Fixes https://github.com/dotnet/roslyn/issues/24112

This uncovered a weird issue in how completion services are used / shared between the normal editor and debugger scenarios. https://github.com/dotnet/roslyn/issues/26394 tracks a fix for that larger issue, while this PR addresses the immediate concern of uninitialized data.

Escrow Template
==================

### Customer scenario

The customer uses completion in the debugger (Watch window, etc.) before ever using completion in the regular editor. Items like `List<>` that are supposed to commit as just `List` will commit as `List<>`, and you then have to go back and clean it up.

### Bugs this fixes

https://github.com/dotnet/roslyn/issues/24112

### Workarounds, if any

Either
1. Run completion in the regular editor one time
2. Allow the incorrect code to be inserted by completion and then go clean it up before proceeding

### Risk

The risk of this change is pretty low. We're making sure a cache is populated when it is first needed.

### Performance impact

Basically none. No normal scenarios are impacted, and in the case of debugger-completion-before-editor-completion, we do a little bit of mef work when the completion list is shown where we didn't before, but that's necessary for correctness.

### Is this a regression from a previous update?

No

### Root cause analysis

The architecture of this partially lazily initialized type relies on an assumption that isn't being met currently. 

### How was the bug found?

Ad hoc testing

### CODE REVIEW

File: src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs [hunk 1]
```diff
@@ -194,7 +194,11 @@ internal protected CompletionProvider GetProvider(CompletionItem item)
             {
                 lock (_gate)
                 {
-                    _nameToProvider.TryGetValue(name, out provider);
+                    if (!_nameToProvider.TryGetValue(name, out provider))
+                    {
+                        CreateRoleProviders(ImmutableHashSet.Create<string>());
```
Comment (rchande) [inline src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs]: I guess this is normally initialized by the call to GetProviders? Why isn't that getting called?

File: src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs [hunk 2]
```diff
@@ -194,7 +194,15 @@ internal protected CompletionProvider GetProvider(CompletionItem item)
             {
                 lock (_gate)
                 {
-                    _nameToProvider.TryGetValue(name, out provider);
+                    if (!_nameToProvider.TryGetValue(name, out provider))
+                    {
+                        // Sometimes the _nameToProvider map hasn't been fully initialized in
+                        // debugger scenarios. If we couldn't find the provider, ensure everything
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs]: hrmm.. that makes me feel like that's an issue in that code path.  How can the debugger sidestep the initialization of the providers?

**Comment by dpoeschl** [comment]
retest WIP please

File: src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs [hunk 3]
```diff
@@ -32,6 +32,29 @@ public abstract class CompletionServiceWithProviders : CompletionService, IEqual
...
+        /// used for showing the Description and for Committing. If you do completion in the 
+        /// debugger before ever doing completion in a regular open document, then the completion
+        /// service being used to commit items hasn't been initialized (that is, 
+        /// ShouldTriggerCompletion has never been called). So, when it's time to lookup the
+        /// provider for the committed item, the cache has not been created, so we don't find
+        /// the provider, and a default provider is used which inserts the DisplayText. This is
+        /// a problem for entries like "List&lt;&gt;", where the insertion text and DisplayText
+        /// differ. 
+        /// Refactoring such that the same completion service is used for triggering and 
+        /// descriptions/committing in debugger scenarios is risky, so we're taking a more
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs]: really?   it seems super strange that we can even get into this situatoin in the first place, and it makes me very quesy about how these services are created and then used.  i mean, it basically hopes/prays htat the one service will have no problem and have no expectations when getting data from another service.  As htis bug indicates, that's clearly not a good assumption to make :)

File: src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs [hunk 4]
```diff
@@ -194,6 +218,12 @@ internal protected CompletionProvider GetProvider(CompletionItem item)
             {
                 lock (_gate)
                 {
+                    if (!_roleProvidersInitialized)
+                    {
+                        CreateRoleProviders(ImmutableHashSet.Create<string>());
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs]: Why is ```ImmutableHashSet.Create<string>()``` the right value to use here?  Can you explain/comment why?

File: src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs [hunk 5]
```diff
@@ -194,6 +218,12 @@ internal protected CompletionProvider GetProvider(CompletionItem item)
...
+                        _roleProvidersInitialized = true;
+                    }
+
                     _nameToProvider.TryGetValue(name, out provider);
```
Comment (sharwell) [inline src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs]: ❓ Why not change this to a `GetOrAdd` call that initializes the requested provider on demand, similar to how `_rolesToProviders` is used above? It seems this would be a clean (non-hacky) way to address the problem.

File: src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs]: Also, still not quite understanding how this won't still have a problem elsewhere. for example, say that the item references a provider that thsi service doesn't know about.  Won't this just return null, causing a crash elsewhere?  or are we resilient to that?

File: src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs [hunk 6]
```diff
@@ -194,13 +196,19 @@ internal protected CompletionProvider GetProvider(CompletionItem item)
...
+                    provider = _nameToProvider.GetOrAdd(name, _getProviderByName);
                 }
             }
 
             return provider;
         }
 
+        private CompletionProvider GetProviderByName(string providerName)
+        {
+            var providers = GetProviders(roles: null);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs]: can a comment be provided explaining why passing null for roles is correct here.  I'm not quite understanding.

**Comment by dpoeschl** [comment]
retest windows_debug_spanish_unit32_prtest please

File: src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs [hunk 7]
```diff
@@ -194,13 +196,19 @@ internal protected CompletionProvider GetProvider(CompletionItem item)
...
+                    provider = _nameToProvider.GetOrAdd(name, _getProviderByName);
                 }
             }
 
             return provider;
         }
 
+        private CompletionProvider GetProviderByName(string providerName)
+        {
+            var providers = GetAllProviders(roles: ImmutableHashSet<string>.Empty);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs]: a comment here explaining why this is the right set of roles would be good.  i actually do not know why this is correct.

File: src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs]: Can this be answered?

File: src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs]: can this be answered?  It seems super weird to me that we even got into this situatoin in the first place.  What codepath it working with multiple CompletionServices... adn why not just fix that codepath?  That seems like teh real fix to have here...

**Comment by dpoeschl** [comment]
Tagging @Jinujoseph for consideration.

**Comment by jinujoseph** [comment]
@dpoeschl  can you retarget this to master (15.8.Preview2 )

**Comment by jinujoseph** [comment]
Approved to merge for 15.8.Preview2 (once we have green test )

**Comment by CyrusNajmabadi** [comment]
It concerns me that none of this has any documentation explaining waht is going on.  This sort of fix invariably leads to problems that get compounded as more and more patches are piled on.
